### PR TITLE
Harrison/gpt4all

### PR DESCRIPTION
**Pull request by hwchase17**

### CODE REVIEW

File: langchain/llms/__init__.py [hunk 1]
```diff
@@ -36,6 +37,7 @@
     "DeepInfra",
     "ForefrontAI",
     "GooseAI",
+    "GPT4All",
```
Comment (eyurtsev) [inline langchain/llms/__init__.py:40]: What do you think about testing __all__ for changes? (unrelated to this PR obviously)

File: langchain/llms/gpt4all.py [hunk 2]
```diff
@@ -0,0 +1,182 @@
...
+
+
+class GPT4All(LLM, BaseModel):
+    r"""Wrapper around GPT4All language models.
+
+    To use, you should have the ``pyllamacpp`` python package installed, the
+    pre-trained model file, and the model's config information.
+
+    Example:
+        .. code-block:: python
```
Comment (eyurtsev) [inline langchain/llms/gpt4all.py:17]: You may need a newline

```suggestion
        .. code-block:: python
        
```

File: langchain/llms/gpt4all.py [hunk 3]
```diff
@@ -0,0 +1,182 @@
...
+            model = GPT4All(model="./models/gpt4all-model.bin", n_ctx=512, n_threads=8)
+
+            # Simplest invocation
+            response = model("Once upon a time, ")
+    """
+
+    model: str
+    """Path to the pre-trained GPT4All model file."""
+
+    n_ctx: int = Field(512, alias="n_ctx")
```
Comment (eyurtsev) [inline langchain/llms/gpt4all.py:28]: Why is alias needed here? I thought it was for reserved fields?

File: tests/integration_tests/llms/test_gpt4all.py [hunk 4]
```diff
@@ -0,0 +1,34 @@
+# flake8: noqa
+"""Test Llama.cpp wrapper."""
+import os
+from urllib.request import urlretrieve
+
+from langchain.llms import GPT4All
+
+
+def get_model() -> str:
```
Comment (eyurtsev) [inline tests/integration_tests/llms/test_gpt4all.py:9]: doesn't really matter 

`save_model_to_file`? `download_model` ?

**Review by eyurtsev — APPROVED** [review]

File: langchain/llms/gpt4all.py [hunk 5]
```diff
@@ -0,0 +1,182 @@
...
+
+        except ImportError:
+            raise ValueError(
+                "Could not import pyllamacpp python package. "
+                "Please install it with `pip install pyllamacpp`."
+            )
+        return values
+
+    @property
+    def _identifying_params(self) -> Mapping[str, Any]:
```
Comment (eyurtsev) [inline langchain/llms/gpt4all.py:143]: Is usage of immutable on return type (and mutables on input like for `stop` token) intentional?

(I tend to flip it around)

File: langchain/llms/gpt4all.py (hunk 3 above)
Comment (hwchase17) [reply langchain/llms/gpt4all.py:29]: @hinthornw ?

File: langchain/llms/__init__.py (hunk 1 above)
Comment (hwchase17) [reply langchain/llms/__init__.py:40]: what do you mean?

File: langchain/llms/gpt4all.py (hunk 5 above)
Comment (hwchase17) [reply langchain/llms/gpt4all.py:144]: not really tbh. i dont think this is done with the intention of being immutable on return type, but rather to expose a helper internal method

File: tests/integration_tests/llms/test_gpt4all.py (hunk 4 above)
Comment (hwchase17) [reply tests/integration_tests/llms/test_gpt4all.py:9]: `_download_model`

**Pull request merged**